db/db_impl/db_impl.cc

```diff
@@ -1572,27 +1567,45 @@ Status DBImpl::ApplyWALToManifest(VersionEdit* synced_wals) {
 Status DBImpl::LockWAL() {
   {
     InstrumentedMutexLock lock(&mutex_);
-    WriteThread::Writer w;
-    write_thread_.EnterUnbatched(&w, &mutex_);
-    WriteThread::Writer nonmem_w;
-    if (two_write_queues_) {
-      nonmem_write_thread_.EnterUnbatched(&nonmem_w, &mutex_);
-    }
+    if (lock_wal_write_token_) {
+      ++lock_wal_extra_outstanding_;
+    } else {
+      WriteThread::Writer w;
+      write_thread_.EnterUnbatched(&w, &mutex_);
+      WriteThread::Writer nonmem_w;
+      if (two_write_queues_) {
+        nonmem_write_thread_.EnterUnbatched(&nonmem_w, &mutex_);
+      }
 
-    lock_wal_write_token_ = write_controller_.GetStopToken();
+      lock_wal_write_token_ = write_controller_.GetStopToken();
 
-    if (two_write_queues_) {
-      nonmem_write_thread_.ExitUnbatched(&nonmem_w);
+      if (two_write_queues_) {
+        nonmem_write_thread_.ExitUnbatched(&nonmem_w);
+      }
+      write_thread_.ExitUnbatched(&w);
```

Comment:
Is there a better way to do this? I suggested using `{Enter,Exit}Unbatched()` originally but this usage looks problematic because

(1) DB mutex can be released during `EnterUnbatched()` so this else branch can be entered by parallel `LockWAL()` callers
(2) [existing] `EnterUnbatched()` will wait for an existing write stall to clear, which seems pointless

Reply:
Well that's odd. Why would we acquire the DB mutex if EnterUnbatched doesn't need the mutex for anything but to release and re-acquire it?

Anyway, easy fix.